### PR TITLE
Implement 2 model specific methods

### PR DESCRIPTION
**Pull request by a-froghyar**
- custom control over the loss dict and the optimizer before the main
optimizer step
- custom model specific implementation for gradient clipping

### CODE REVIEW

**Comment by a-froghyar** [comment]
After rebasing my TTS PR, I've realised that the optimizers have also been outsourced to this repo, so this PR also implements a `StepwiseGradualLR` scheduler. This is a scheduler used to hardcode step-lr pairs in the config the following way:

```
 lr_scheduler="StepwiseGradualLR",
    lr_scheduler_params={
        "gradual_learning_rates": [
            [0, 1e-3],
            [2e4, 5e-4],
            [4e5, 3e-4],
            [6e4, 1e-4],
            [8e4, 5e-5],
        ]
    },
```

File: trainer/trainer.py [hunk 1]
```diff
@@ -991,12 +991,19 @@ def _optimize(
                         loss_dict["amp_scaler"] = scaler.get_scale()  # for logging
                     update_lr_scheduler = scale_prev <= scaler.get_scale()
         else:
+            if hasattr(self.model, "before_main_optimizer_step"):
+                # model dependent control for custom backward and optimizer operations
+                # before the main optimizer step
+                self.model.before_main_optimizer_step(loss_dict, optimizer)
```
Comment (erogol) [inline trainer/trainer.py:997]: This line should use ```TrainerCallbacks``` in callbacks.py to be widely reachable.

**Review by erogol — CHANGES_REQUESTED** [review]

File: trainer/trainer.py (hunk 1 above)
Comment (a-froghyar) [reply trainer/trainer.py:997]: @erogol  Is it ok to pass in `loss_dict` and `optimizer` to a callback?

Comment (a-froghyar) [reply trainer/trainer.py]: @erogol pushed up the callback, please check

File: trainer/torch.py [hunk 2]
```diff
@@ -68,3 +82,36 @@ def get_lr(self):
             base_lr * self.warmup_steps**0.5 * min(step * self.warmup_steps**-1.5, step**-0.5)
             for base_lr in self.base_lrs
         ]
+
+
+# pylint: disable=protected-access
+class StepwiseGradualLR(torch.optim.lr_scheduler._LRScheduler):
```
[the remark this answers is not in the document]
Comment (a-froghyar) [reply trainer/torch.py:88]: @WeberJulian is helping on this, thank you!

**Comment by a-froghyar** [comment]
Ready from my side! Thank you @WeberJulian for the help

File: trainer/callbacks.py [hunk 3]
```diff
@@ -59,6 +59,15 @@ def on_epoch_end(trainer) -> None:
         if hasattr(trainer.optimizer, "on_epoch_end"):
             trainer.optimizer.on_epoch_end(trainer)
 
+    @staticmethod
+    def before_main_optimizer_step(trainer, loss_dict, optimizer):
```
Comment (erogol) [inline trainer/callbacks.py:63]: why do you need to pass loss_dict and optimizer? They are already reachable by the trainer.

File: trainer/trainer.py [hunk 4]
```diff
@@ -1014,12 +1014,16 @@ def _optimize(
                         loss_dict["amp_scaler"] = scaler.get_scale()  # for logging
                     update_lr_scheduler = scale_prev <= scaler.get_scale()
         else:
+            self.callbacks.before_main_optimizer_step(self, loss_dict, optimizer)
```
Comment (erogol) [inline trainer/trainer.py:1017]: I'd call this "before_backward_pass" 

and move ```self.model.apply_gradient_clipping``` in ```before_gradient_clipping```

**Review by erogol — CHANGES_REQUESTED** [review]

File: trainer/callbacks.py (hunk 3 above)
Comment (a-froghyar) [reply trainer/callbacks.py:63]: I need to have access to them in the model method --> https://github.com/coqui-ai/TTS/blob/1afd685d230a70b9651d67392b5096dd76042ceb/TTS/tts/models/tacotron2.py#L303

File: trainer/trainer.py (hunk 4 above)
Comment (a-froghyar) [reply trainer/trainer.py:1017]: @erogol  do you mean wrapping `self.model.apply_gradient_clipping` into a separate callback?

**Comment by WeberJulian** [comment]
> why do you need to pass loss_dict and optimizer? They are already reachable by the trainer.

> I need to have access to them in the model method --> https://github.com/coqui-ai/TTS/blob/1afd685d230a70b9651d67392b5096dd76042ceb/TTS/tts/models/tacotron2.py#L303

He means that since the first argument of the callback is trainer, in this function, you can call trainer.optimizer, no need to pass it as argument.

Look at how callbacks are implemented in TTS

**Comment by WeberJulian** [comment]
> I'd call this "before_backward_pass"

> and move self.model.apply_gradient_clipping in before_gradient_clipping

> @erogol do you mean wrapping self.model.apply_gradient_clipping into a separate callback?

No he means move the content of `self.model.apply_gradient_clipping` inside a new callback called `before_gradient_clipping` and implement this callback in your model instead of the `apply_gradient_clipping` function.

**Comment by a-froghyar** [comment]
Changes implemented in both repos!

**Comment by erogol** [comment]
I've merged this after rebasing. Thanks for the PR @a-froghyar
